<a id='top'></a>

# CSCI 3022: Intro to Data Science - Spring 2019 Practicum 2
***

This practicum is due on Canvas by **11:59 PM on Friday May 3**. Your solutions to theoretical questions should be done in Markdown/MathJax directly below the associated question.  Your solutions to computational questions should include any specified Python code and results as well as written commentary on your conclusions.  

**Here are the rules:** 

1. All work, code and analysis, must be your own. 
1. You may use your course notes, posted lecture slides, textbooks, in-class notebooks, and homework solutions as resources.  You may also search online for answers to general knowledge questions like the form of a probability distribution function or how to perform a particular operation in Python/Pandas. 
1. This is meant to be like a coding portion of your midterm exam. So, the instructional team will be much less helpful than we typically are with homework. For example, we will not check answers, help debug your code, and so on.
1. If something is left open-ended, it is because we want to see how you approach the kinds of problems you will encounter in the wild, where it will not always be clear what sort of tests/methods should be applied. Feel free to ask clarifying questions though.
2. You may **NOT** post to message boards or other online resources asking for help.
3. You may **NOT** copy-paste solutions *from anywhere*.
4. You may **NOT** collaborate with classmates or anyone else.
5. In short, **your work must be your own**. It really is that simple.

Violation of the above rules will result in an immediate academic sanction (*at the very least*, you will receive a 0 on this practicum or an F in the course, depending on severity), and a trip to the Honor Code Council.

**By submitting this assignment, you agree to abide by the rules given above.**

***

**Name**:  Chakrya Ros

***


**NOTES**: 

- You may not use late days on the practicums nor can you drop your practicum grades. 
- If you have a question for us, post it as a **PRIVATE** message on Piazza.  If we decide that the question is appropriate for the entire class, then we will add it to a Practicum clarifications thread. 
- Do **NOT** load or use any Python packages that are not available in Anaconda 3.6. 
- Some problems with code may be autograded.  If we provide a function API **do not** change it.  If we do not provide a function API then you're free to structure your code however you like. 
- Submit only this Jupyter notebook to Canvas.  Do not compress it using tar, rar, zip, etc. 
- This should go without saying, but... For any question that asks you to calculate something, you **must show all work to receive credit**. Sparse or nonexistent work will receive sparse or nonexistent credit.

---
**Shortcuts:**  [Problem 1](#p1) | [Problem 2](#p2) | [Bottom](#bot)

---

In [2]:
# import whatever packages you think you will need here!
from scipy import stats
import numpy as np 
import pandas as pd
import statsmodels.api as sm
import matplotlib.pyplot as plt
%matplotlib inline


<br>

---
<a id='p1'></a>
[Back to top](#top)

### [30 points] Problem 1:  What's the best Oreo flavor?

> <img src="https://media2.s-nbcnews.com/j/newscms/2015_52/1351976/151223-watermelon-oreos-flickr-yh-1234p_a4c4ba1016261e5d799a789666948a6e.fit-760w.jpg" style="width: 400px;"/>
**Figure 1.** Watermelon Oreos. These are undeniably the absolute _worst_ Oreo flavor, but not a part of our data set. If they were, the problem would be too easy.

You're skipping down the road one day, singing a merry tune without a care in the world. Every so often you hop and click your heels together. After one such hop, though, you stumble and nearly trip. Under your feet is a fun data set!

In particular, this rascally data set includes ratings from 5 different individuals for 12 different types of Oreo cookie. Because some of the raters are allergic to disgusting flavors of Oreo, not every taster tasted every type of cookie. More information about the data set can be found [here](https://www.kaggle.com/rtatman/oreo-flavors-tastetest-ratings).

As you examine the data, a booming voice from the sky instructs you to analyze the data to answer two key questions:
1. is there some difference in mean rating given to different flavors of Oreo? and
2. which flavor (out of those in the data set) is the best?

**Part A:** Load the data set as a Pandas DataFrame. Then, process it as you see fit. For the analysis, you might create a new DataFrame, or other kinds of Python objects to use the data. Do what feels right, and include an **explanation** in Markdown of how you are processing the data. You **may not** alter the original data file.

In [1]:
pf = pd.read_csv("oreo_rankings.csv")
pf.head()

NameError: name 'pd' is not defined

**Part B:** In the remainder of this problem, you will perform some hypothesis tests to examine whether these data suggest there are significant differences in the mean ratings of the different Oreo flavors. Pick a level of significance for these experiments, and explain how you decided to use that significance level. "Because we used it a lot in class" is ***not*** a good reason.

**Part C:** Perform an **Analysis of Variance hypothesis test** in order to determine if there is evidence that there is _some_ difference among the mean ratings given to these 12 Oreo flavors. Clearly state your null and alternative hypothesis, and use the significance level identified in **Part B**. You must show **all** calculations **by hand** (and may of course use Python as a calculator, and to compute values from a distribution using the appropriate percent-point-function (ppf) or cumulative distribution function (cdf)).

In addition to showing the code for your calculations, make comments **in Markdown** explaining what you are doing.

In [ ]:
dfnew1 = df.copy()
dfnew1 = dfnew1.replace({"-":np.nan})
dfOreo = pd.DataFrame(columns=['pb&j', 'mega stuf', "lemon", "chocolate","birthday cake","double stuf",
                               "red velvet","Dunkin' Donuts Mocha","mini","thins, coconut","mint","cinnamon bun"])

list_ = {0:'pb&j', 1:'mega stuf', 2:"lemon", 3:"chocolate",4:"birthday cake",5:"double stuf",
            6:"red velvet",7:"Dunkin' Donuts Mocha",8:"mini",9:"thins, coconut",10:"mint",11:"cinnamon bun"}

def newDataFrame(df, list_):
    df_new = pd.DataFrame(columns=['pb&j', 'mega stuf', "lemon", "chocolate","birthday cake","double stuf",
                               "red velvet","Dunkin' Donuts Mocha","mini","thins, coconut","mint","cinnamon bun"])
    
    for l in list_:
        x = df.loc[df['oreo_flavor'] == list_[l], 'rater_n':'rater_a']
        x = pd.to_numeric(x.loc[l])
        df_new[list_[l]] = x
    return df_new

dfOreo = newDataFrame(dfnew1, list_)
# dfOreo = dfOreo.dropna()
dfOreo

**Part D:** Perform a **t hypothesis test** to determine if there is evidence supporting the claim that cinnamon bun Oreos have a higher mean rating than PB&J Oreos. Use the significance level you identified in **Part B**. Clearly state your null and alternative hypotheses, your conclusions, and show all work. Again, you may not use any canned t-test function.

In [ ]:
n = 7
xbarCin = dfOreo['cinnamon bun'].mean()
varCin = np.sum((dfOreo['cinnamon bun'] - xbarCin)**2)/n-1

xbarPB = dfOreo['pb&j'].mean()
varPB = np.sum((dfOreo['pb&j'] - xbarPB)**2)/n-1
# print(varPB)
# print(varCin)

numerator = xbarCin - xbarPB
denomerator = np.sqrt((varCin/n) + (varPB/n))
T = numerator / denomerator
# print(numerator)
# print(denomerator)

print("Test_statistic T = {:.3f}".format(T))
p_value =stats.t.cdf(T,22)
print("p-value = {:.3f}".format(p_value))
if T < t_alp :
    print("Reject Null Hypothesis")
else:
    print("Failed to reject Null")

**Part E:** Do your results from Parts C and D agree with one another? If they agree, explain how they are in agreement in _words_. If they do not agree, explain why you think they do not agree.

_Hint: if they do not agree, consider carefully what assumptions we make in ANOVA and the t-test._

<br>

---
<a id='p2'></a>
[Back to top](#top)

### [30 points] Problem 2: Multiple Linear Regression to Explain House Hauntings

<img src="https://s-media-cache-ak0.pinimg.com/originals/09/72/01/09720128cff5de4d4af038cd3fcf7f69.jpg" style="width: 300px;"/>

In an effort to control the skyrocketing prices of real estate in the Colorado Front Range, Governor Polis implemented a cutting edge new intervention. This new program oversaw the introduction of ghosts back into their natural ecosystem, after the ghost population seriously dwindled in recent decades due to overhaunting. However, an unfortunate miscalculation has led to haunted houses becoming a very serious problem in Colorado. Modern problems require modern solutions, so Governor Polis has hired you and the famous hedgehog data scientist/part-time ghostbuster Amy to determine what features of a house may be used to best predict a `haunted` score, related to the probability that a house with the given features is haunted (higher $\leftrightarrow$ more likely to be haunted).

You decide to use multiple linear regression to understand and predict what factors lead to increased haunted house hazard. You collected a data set from Haunted Zillow, the lesser-known database of haunted house prices and attributes. The data cover a variety of potential features, and you'll find this data in the file `houses.csv`. 

**Response**: 

- $\texttt{haunted}$: a haunting score, related to the probability that a house with the given features is haunted (higher $\leftrightarrow$ more likely to be haunted)

**Features**: 

- $\texttt{age}$: age of the house, in years
- $\texttt{area}$: square footage of interior of house
- $\texttt{bathrooms}$: number of bathrooms
- $\texttt{distance metro}$: distance to the nearest major metropolitan area (in miles)
- $\texttt{distance cemetery}$: distance to the nearest cemetery (in miles)
- $\texttt{cats}$: the number of cats within a one-block radius of the house
- $\texttt{howls}$: the number of wolf howls heard on an average night in the house's neighborhood
- $\texttt{clouds}$: what percentage of the sky was covered by clouds (fraction, 0-1)
- $\texttt{precipitation}$: amount of precipitation in the past 72 hours (inches)
- $\texttt{misery index}$: an economic indicator for how miserable the average United States citizen is, based on the unemployment rate and the inflation rate. More [here](https://www.stuffyoushouldknow.com/podcasts/whats-the-misery-index.htm) and [here](https://en.wikipedia.org/wiki/Misery_index_(economics)). Higher values correspond to more miserable citizens.
- $\texttt{ice cream sold}$: the number of units of ice cream sold at the farmer's market the week the house was most recently sold

**Part A**: Read the data from `houses.csv` into a Pandas DataFrame.  Note that since we will be doing a multiple linear regression we will need all of the features, so you should drop any row in the DataFrame that is missing data. 

In [4]:
dfHouse = pd.read_csv("houses.csv")
dfHouse = dfHouse.dropna()
dfHouse.head()

age  area  bathrooms  distance metro  distance cemetery  cats  howls  \
0   65.06  2041        1.0             7.1              10.01     7      3   
1  141.48  1564        0.0             7.4               4.07     5      5   
2    7.58  1637        3.0             7.0               3.36     2      0   
3   51.47  2021        2.0             7.9               3.43     6      8   
4  259.44  1642        1.0             7.5               3.19     4      1   

   clouds  precipitation  misery index  ice cream sold   haunted  
0    1.00           0.82         12.99             273 -0.596150  
1    1.00           0.99         16.77             184 -0.146465  
2    1.00           1.17         16.49             141 -0.303117  
3    0.13           0.92          8.28             146  0.339912  
4    1.00           1.73          5.90             178  0.724867

**Part B**: Perform the appropriate statistical test at the $\alpha = 0.01$ significance level to determine if _at least one_ of the features is related to the the response $y$.  Clearly describe your methodology and show all computations in Python. 

**Solution:**

If there is a significant linear relationship between the independent variable X and the dependent variable Y, the slope will not equal zero.

Hypotheses testing:
$$
H_0 : \beta = 0 \\
H_1 : \beta \ne 0
$$

Test Statistic:

$$ TB = \dfrac{\hat{\beta} - 0}{SE(\hat{\beta})}$$

where, $SE(\hat{\beta}) = \dfrac{\hat{\sigma}}{\sqrt{\sum(x_i - \bar{x})^2}}$

and $$\hat{\sigma} = \sqrt{\dfrac{SSE}{n-2}} = \sqrt{\dfrac{\sum(y_i - \hat{y}_i)^2)}{n-2}} $$

**p-value** = 2*(stats.t.cdf(T, df=len(x)-2) < $\alpha$

Is there sufficient evidence to conclude that the observed relationship between the features is related to the response y.

The features that we want to test are age, area, bathrooms.

In [11]:
def findPvalue(x, y, name_features):
    print(name_features)
    bhat, ahat, rval, pval, stderr = stats.linregress(x, y)
    yhat = ahat + bhat*x
    SSE = np.sum((y - yhat)**2)
    print("SSE : {}".format(SSE))
    varhat = SSE/ (len(x)-2)

    xbar = np.mean(x)
    SEbeta = np.sqrt(varhat/np.sum((x-xbar)**2))
    TB = (bhat-0)/SEbeta
    tpval = 2*(stats.t.cdf(TB, df=len(x)-2))
    return tpval

In [12]:
#Age
x_age = dfHouse["age"]
y_age = dfHouse["age"]
pval_age = findPvalue(x_age, y_age, "age")
print("p-value = {}".format(pval_age))

age
SSE : 0.0
p-value = 2.0


C:\Users\chakrya\Anaconda3\lib\site-packages\ipykernel_launcher.py:11: RuntimeWarning: divide by zero encountered in double_scalars
  # This is added back by InteractiveShellApp.init_path()


In [13]:
#Area
x_area = dfHouse["area"]
y_area = dfHouse["haunted"]
pval_area = findPvalue(x_area, y_area, "Area")
print("p-value = {}".format(pval_area))

Area
SSE : 52.4748766215231
p-value = 1.1045346795380375


In [16]:
#bathroom
x_area = dfHouse["bathrooms"]
y_area = dfHouse["haunted"]
pval_area = findPvalue(x_area, y_area, "bathrroms")
print("p-value = {}".format(pval_area))

bathrroms
SSE : 51.91419114115229
p-value = 0.3923476025816687


Since the p-value is greater than  𝛼=0.01  we fail to reject the null hypothesis and conclude that there is not sufficient statistical evidence that  $\beta \ne 0$  and the relationship between the area and haunted is real.

In [14]:
#distance cemetery
x_bath = dfHouse["distance cemetery"]
pval_bathrooms = findPvalue(x_bath, dfHouse["haunted"], "Bathrooms")
print("p-value = {}".format(pval_bathrooms))

Bathrooms
SSE : 5.788952699088558
p-value = 8.64632347094066e-34


In [17]:
#distance metro
x_area = dfHouse["distance metro"]
y_area = dfHouse["haunted"]
pval_area = findPvalue(x_area, y_area, "distance metro")
print("p-value = {}".format(pval_area))

distance metro
SSE : 51.678187961785646
p-value = 0.3090649388449295


In [18]:
#cats	
x_area = dfHouse["cats"]
y_area = dfHouse["haunted"]
pval_area = findPvalue(x_area, y_area, "cats")
print("p-value = {}".format(pval_area))

cats
SSE : 52.481706967455985
p-value = 1.0739220781179775


In [20]:
#distance metro
x_hows = dfHouse["howls"]
y_hows = dfHouse["haunted"]
pval_hows = findPvalue(x_hows, y_hows, "howls")
print("p-value = {}".format(pval_area))

howls
SSE : 52.486682461392
p-value = 1.0382825706826384


In [21]:
#distance metro
x_area = dfHouse["clouds"]
y_area = dfHouse["haunted"]
pval_area = findPvalue(x_area, y_area, "clounds")
print("p-value = {}".format(pval_area))

clounds
SSE : 51.643991320493015
p-value = 0.29899109096359555


In [23]:
dfHouse.columns

Index(['age', 'area', 'bathrooms', 'distance metro', 'distance cemetery',
       'cats', 'howls', 'clouds', 'precipitation', 'misery index',
       'ice cream sold', 'haunted'],
      dtype='object')

In [24]:
#distance metro
x_area = dfHouse["precipitation"]
y_area = dfHouse["haunted"]
pval_area = findPvalue(x_area, y_area, "precipitation")
print("p-value = {}".format(pval_area))

precipitation
SSE : 52.031738862436605
p-value = 0.44582507382502223


In [25]:
#distance metro
x_area = dfHouse["misery index"]
y_area = dfHouse["haunted"]
pval_area = findPvalue(x_area, y_area, "misery index")
print("p-value = {}".format(pval_area))

misery index
SSE : 52.48740494108533
p-value = 1.0297239779454452


In [26]:
#distance metro
x_area = dfHouse["ice cream sold"]
y_area = dfHouse["haunted"]
pval_area = findPvalue(x_area, y_area, "ice cream sold")
print("p-value = {}".format(pval_area))

ice cream sold
SSE : 50.89268150635988
p-value = 0.1518782750829303


ice cream, clouds, distance metreo, distance ceremen is lowes, bathroom

In [9]:
y = dfHouse["haunted"]
Xfull = dfHouse.loc[:,dfHouse.columns != "haunted"]
Xfull = sm.add_constant(Xfull)
full_model = sm.OLS(y, Xfull).fit()
full_model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                haunted   R-squared:                       0.948
Model:                            OLS   Adj. R-squared:                  0.938
Method:                 Least Squares   F-statistic:                     94.60
Date:                Sun, 21 Apr 2019   Prob (F-statistic):           1.97e-32
Time:                        17:09:12   Log-Likelihood:                 13.575
No. Observations:                  69   AIC:                            -3.151
Df Residuals:                      57   BIC:                             23.66
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                 0.3480      0.323      1.078      0.285      -0.298       0.994
age                   0.0017      0.000      5.859      0.000       0.001       0.002
area                5.83e-05   6.16e-05      0.946      0.348   -6.51e-05       0.000
bathrooms            -0.0411      0.023     -1.806      0.076      -0.087       0.004
distance metro       -0.0142      0.032     -0.443      0.660      -0.079       0.050
distance cemetery    -0.1000      0.003    -30.867      0.000      -0.106      -0.093
cats                  0.0316      0.010      3.258      0.002       0.012       0.051
howls                 0.0056      0.011      0.494      0.623      -0.017       0.029
clouds                0.0902      0.123      0.734      0.466      -0.156       0.336
precipitation         0.0068      0.067      0.102      0.919      -0.128       0.141
misery index         -0.0082      0.006     -1.456      0.151      -0.020       0.003
ice cream sold       -0.0006      0.001     -1.274      0.208      -0.002       0.000
==============================================================================
Omnibus:                        1.629   Durbin-Watson:                   2.191
Prob(Omnibus):                  0.443   Jarque-Bera (JB):                1.480
Skew:                           0.230   Prob(JB):                        0.477
Kurtosis:                       2.450   Cond. No.                     2.55e+04
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.55e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

**Conclusion:**

From the model summery, we see F = 94.60 and p-value $1.94 \times 10^{-32}$.
Since the p-value is less than 𝛼=0.01 we reject the null hypothesis and conclude that there is the sufficient statistical evidence that $\beta \ne 0$ and the relationship between  the  the features (distance cemetery) is related to the response y.

**Part C**: Write a function `forward_select(df, resp_str, maxk)` that takes in the DataFrame, the name of the column corresponding to the response, and the maximum number of desired features, and returns a list of feature names corresponding to the `maxk` most important features via forward selection.  At each stage in forward selection you should add the feature whose inclusion in the model would result in the lowest sum of squared errors $(SSE)$. Use your function to determine the best $k=5$ features to include in the model. Clearly indicate which feature was added in each stage. 

**Note**: The point of this exercise is to see if you can implement **foward_select** yourself.  You may of course use canned routines like statmodels OLS, but you may not call any Python method that explicitly performs forward selection.

In [ ]:
def forward_select(df, resp_str, maxk):
    
    #get respone from dataframe
    y = df[resp_str]
    
    #get features
    features = list(df.columns[df.columns != resp_str])
#     print(features)

forward_select(dfHouse, "haunted", 200)

In [ ]:
y = dfHouse["haunted"]
x = dfhouse.loc[:, ]

In [ ]:
y = dfS["sharknado hazard"]
Xred = dfS.loc[:, good_features]
Xred = sm.add_constant(Xred)
red_model = sm.OLS(y, Xred).fit()
red_model.summary()

**Part D**: Write down the multiple linear regression model, including estimated parameters, obtained by your forward selection process. 

**Part E**: Perform the appropriate statistical test at the $\alpha = 0.05$ significance level to determine whether there is a statistically significant difference between the full model with all features and the reduced model obtained by forward selection in **Part D**. 

**Part F**: Based on your conclusions in **Part E**, use the _better_ of the two models to predict the sharknado hazard when the following features are observed: 

- $\texttt{age}$: 100 years
- $\texttt{area}$: 2200 square feet
- $\texttt{bathrooms}$: 3 bathrooms
- $\texttt{distance metro}$: 25 miles
- $\texttt{distance cemetery}$: 10 miles
- $\texttt{cats}$: 4 cats
- $\texttt{howls}$: 5 wolf howls/night
- $\texttt{clouds}$: 0.65 cloud cover
- $\texttt{precipitation}$: 0 inches
- $\texttt{misery index}$: 10
- $\texttt{ice cream sold}$: 125

**Part G:** Governor Polis dabbles a bit in the art of data science, as well as the science of data art. He tells you that the response (`haunted` score) that you and Amy predicted is actually the natural logarithm of the _odds_ that a house with the given features is haunted, where if $p$ is the probability that a house is haunted, then the odds are given by $$\text{odds} = \dfrac{p}{1-p}$$

What transformation must you make in order to turn your multiple regression model into a logistic regression model, to classify a house a haunted or not? Perform this simple transformation, then use a decision threshold of 0.5 to classify the house from **Part F** as haunted or not haunted. No new models should be fit here; use the same model that you used in Part F.

<br>

---
[Back to top](#top)
<a id='bot'></a>